File: packages/playground/src/option-chain.ts
/**
 * Writes AAPL's near-dated chain — one record per contract, both types, carrying the
 * greeks, the implied volatility and the quote — to `viz/data/option-chain.json`, which
 * `viz/option_chain.py` draws as eight contour panels.
 *
 *   npm run option-chain -w @fleece/playground
 *   uv run --project viz viz/option_chain.py
 *
 * Nothing is filtered: every contract Alpaca could say anything about is written, both
 * sides of spot, with each measure null where it is missing rather than absent or zeroed —
 * a zero delta is a real value a far out-of-the-money call has. One record per contract
 * rather than a grid, because expirations do not share a strike set; the renderer
 * triangulates the scattered points.
 */
import { mkdirSync, writeFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';

import { AlpacaMarketDataClient, type OptionChainRequest, type OptionSnapshot, type OptionType } from '@fleece/marketdata';
import { LoggerFactory, easternClock } from '@fleece/utilities';

import { marketDataKeys } from './credentials';

const logger = LoggerFactory.getLogger('OptionChain');

const UNDERLYING = 'AMZN';

/** Past a few months the term structure flattens and delta is a wall of near-1. */
const EXPIRATION_WINDOW_DAYS = 50;

/**
 * Strikes within this fraction of spot. Widen it for more of the skew, which is only
 * visible in the wings; past about a third out the bid is a penny on a contract nobody has
 * traded, and the volatility solved from it says more about the tick size than the market.
 */
const STRIKE_BAND = 0.15;

/**
 * How far apart the two sides of a strike may sit before the chart says so, as a fraction —
 * `0.01` is one volatility point.
 *
 * Put-call parity says a call and a put on one strike share an implied volatility, so two
 * different numbers mean the greeks and the quotes are marked against different underlying
 * prices. That is the ordinary state after hours, when OPRA has stopped quoting and the
 * stock trades on. Past this line the chain is still written, and the figure is stamped.
 */
const SEAM_TOLERANCE = 0.01;

/**
 * The smallest bid a seam reading will be taken from, in dollars. Only the seam — Alpaca
 * solves a volatility even for a 0.00 x 0.05 quote, and parity checked against tick noise
 * reports a rounding as a disagreement.
 */
const SEAM_MINIMUM_BID = 0.05;

/** Alpaca's page maximum for a chain. */
const PAGE_SIZE = 1_000;

/** `dist/` at runtime, so three levels up is the repo root. */
const ROOT = resolve(__dirname, '../../..');
const OUTPUT = resolve(ROOT, 'viz/data/option-chain.json');

/**
 * One contract, flat. Every measure is nullable and every key always present, so a missing
 * field never has to be told apart from a zero — Alpaca omits the greeks and the
 * volatility for roughly four contracts in ten of a large chain.
 */
interface ContractRecord {
  readonly symbol: string;
  readonly expiration: string;
  readonly strike: number;
  readonly type: OptionType;
  readonly delta: number | null;
  readonly gamma: number | null;
  readonly theta: number | null;
  readonly vega: number | null;
  readonly rho: number | null;
  /** As a fraction: `0.31` is 31%. */
  readonly iv: number | null;
  readonly bid: number | null;
  readonly ask: number | null;
  /** The middle of the quote, in dollars per share. */
  readonly mid: number | null;
}

/** One expiration's disagreement between its two sides, at the strike nearest spot. */
interface SeamReading {
  readonly expiration: string;
  /** `|call iv - put iv|`, as a fraction. */
  readonly disagreement: number;
  /** How far the underlying behind the greeks sits above the one behind the quotes, in dollars. */
  readonly greeksAhead: number;
}

interface Seam {
  /** How many expirations carried a strike quoted on both sides. */
  readonly expirations: number;
  /** Median across them, as a fraction. */
  readonly disagreement: number;
  /** Median price error behind it, in dollars. Positive means the greeks lead. */
  readonly greeksAhead: number;
  readonly tolerance: number;
  readonly stale: boolean;
}

/**
 * The last close, as the centre of the strike band. A week back rather than today, because
 * `dailyBars` answers about sessions and a one-day window is empty before the open.
 */
async function lastClose(client: AlpacaMarketDataClient, today: string): Promise<number> {
  const { bars } = await client.dailyBars({ symbol: UNDERLYING, from: easternClock.shiftDate(today, -7), to: today });
  const latest = bars.at(-1);
  if (latest === undefined) {
    throw new Error(`Alpaca has no daily bar for ${UNDERLYING} in the last week, so there is no spot price to centre the strikes on.`);
  }
  return latest.c;
}

/** Every page of the chain. The client hands back one page and a cursor; walking it is the caller's job. */
async function wholeChain(client: AlpacaMarketDataClient, request: OptionChainRequest): Promise<ReadonlyArray<OptionSnapshot>> {
  const snapshots: OptionSnapshot[] = [];
  let startAfter: string | undefined = undefined;

  do {
    const page = await client.optionChain({ ...request, limit: PAGE_SIZE, startAfter });
    snapshots.push(...page.contracts);
    startAfter = page.resumeFrom;
  } while (startAfter !== undefined);

  return snapshots;
}

/** The middle of the quote. An ask of zero means no quote at all, not a free option. */
function midOf(snapshot: OptionSnapshot): number | null {
  const { lq } = snapshot;
  if (lq === undefined || lq.ap <= 0) {
    return null;
  }
  return (lq.bp + lq.ap) / 2;
}

/**
 * The seam, one reading per expiration, at the strike nearest spot — where vega is largest
 * and so least disturbed by a penny of rounding in either quote.
 */
function readSeams(snapshots: ReadonlyArray<OptionSnapshot>, spot: number): ReadonlyArray<SeamReading> {
  const sides = new Map<string, { call?: OptionSnapshot; put?: OptionSnapshot }>();

  for (const snapshot of snapshots) {
    const { iv, greeks, lq, contract } = snapshot;
    if (typeof iv !== 'number' || greeks === undefined || lq === undefined || lq.bp < SEAM_MINIMUM_BID) {
      continue;
    }
    const key = `${contract.expiration}:${contract.strike}`;
    const pair = sides.get(key) ?? {};
    sides.set(key, contract.type === 'call' ? { ...pair, call: snapshot } : { ...pair, put: snapshot });
  }

  const nearest = new Map<string, SeamReading & { distance: number }>();
  for (const { call, put } of sides.values()) {
    if (call?.iv === undefined || put?.iv === undefined || call.greeks === undefined) {
      continue;
    }
    const { expiration, strike } = call.contract;
    const distance = Math.abs(strike - spot);
    const previous = nearest.get(expiration);
    if (previous !== undefined && previous.distance <= distance) {
      continue;
    }
    nearest.set(expiration, {
      expiration,
      distance,
      disagreement: Math.abs(call.iv - put.iv),
      greeksAhead: (put.iv - call.iv) * 100 * call.greeks.vega,
    });
  }

  return [...nearest.values()].map(({ expiration, disagreement, greeksAhead }) => ({ expiration, disagreement, greeksAhead }));
}

/** The middle reading, so one ragged expiration cannot set the stamp. */
function median(values: ReadonlyArray<number>): number {
  const sorted = [...values].sort((a, b) => a - b);
  const middle = Math.floor(sorted.length / 2);
  return sorted.length % 2 === 0 ? (sorted[middle - 1]! + sorted[middle]!) / 2 : sorted[middle]!;
}

/** Null when no strike is quoted on both sides. */
function measureSeam(snapshots: ReadonlyArray<OptionSnapshot>, spot: number): Seam | null {
  const seams = readSeams(snapshots, spot);
  if (seams.length === 0) {
    return null;
  }

  const disagreement = median(seams.map((seam) => seam.disagreement));
  return {
    expirations: seams.length,
    disagreement,
    greeksAhead: median(seams.map((seam) => seam.greeksAhead)),
    tolerance: SEAM_TOLERANCE,
    stale: disagreement > SEAM_TOLERANCE,
  };
}

/**
 * Every contract Alpaca said something about, oldest expiration first. Only one carrying
 * neither greeks nor a quote is dropped, since no panel could draw it.
 */
function flatten(snapshots: ReadonlyArray<OptionSnapshot>): { contracts: ReadonlyArray<ContractRecord>; empty: number } {
  const contracts: ContractRecord[] = [];
  let empty = 0;

  for (const snapshot of snapshots) {
    const { greeks, contract, lq } = snapshot;
    const mid = midOf(snapshot);

    if (greeks === undefined && mid === null) {
      empty += 1;
      continue;
    }

    contracts.push({
      symbol: snapshot.S,
      expiration: contract.expiration,
      strike: contract.strike,
      type: contract.type,
      delta: greeks?.delta ?? null,
      gamma: greeks?.gamma ?? null,
      theta: greeks?.theta ?? null,
      vega: greeks?.vega ?? null,
      rho: greeks?.rho ?? null,
      iv: snapshot.iv ?? null,
      bid: lq?.bp ?? null,
      ask: lq?.ap ?? null,
      mid,
    });
  }

  contracts.sort((a, b) => {
    if (a.expiration !== b.expiration) {
      return a.expiration.localeCompare(b.expiration);
    }
    return a.strike === b.strike ? a.type.localeCompare(b.type) : a.strike - b.strike;
  });

  return { contracts, empty };
}

/** The seam in a sentence, for the log. */
function describeSeam(seam: Seam | null, spot: number): string {
  if (seam === null) {
    return `No strike is quoted on both sides, so put-call parity could not be checked against ${UNDERLYING}'s quotes.`;
  }
  if (!seam.stale) {
    return `Puts and calls agree to ${(seam.disagreement * 100).toFixed(2)} volatility points at the money across ${seam.expirations} expirations.`;
  }
  return [
    `Puts and calls disagree by ${(seam.disagreement * 100).toFixed(1)} volatility points at the money, against a tolerance of ${(seam.tolerance * 100).toFixed(1)}.`,
    `Parity says a strike has one volatility, so this is not skew — it is the greeks being marked against an underlying about ${Math.abs(seam.greeksAhead).toFixed(2)} dollars`,
    `${seam.greeksAhead >= 0 ? 'above' : 'below'} the ${spot.toFixed(2)} the quotes were struck at.`,
    'That is what the evening looks like: OPRA stops quoting at the option close and the stock trades on without it.',
    'The call and put rows will sit at different levels; run this during the option session and they come back into step.',
  ].join(' ');
}

async function main(): Promise<void> {
  const client = new AlpacaMarketDataClient(marketDataKeys());
  const today = easternClock.date();
  const spot = await lastClose(client, today);
  const expirationTo = easternClock.shiftDate(today, EXPIRATION_WINDOW_DAYS);

  logger.info(`${UNDERLYING} last closed at ${spot.toFixed(2)}. Fetching calls and puts expiring ${today} to ${expirationTo}, within ${STRIKE_BAND * 100}% of that.`);

  const snapshots = await wholeChain(client, {
    underlying: UNDERLYING,
    expirationFrom: today,
    expirationTo,
    strikeFrom: Math.round(spot * (1 - STRIKE_BAND)),
    strikeTo: Math.round(spot * (1 + STRIKE_BAND)),
  });

  const seam = measureSeam(snapshots, spot);
  const note = describeSeam(seam, spot);
  if (seam?.stale ?? true) {
    logger.warn(note);
  } else {
    logger.info(note);
  }

  const { contracts, empty } = flatten(snapshots);
  if (contracts.length === 0) {
    throw new Error(`Alpaca priced none of the ${snapshots.length} contracts in the band. There is nothing to plot.`);
  }

  const expirations = new Set(contracts.map((contract) => contract.expiration));
  const calls = contracts.filter((contract) => contract.type === 'call').length;

  mkdirSync(dirname(OUTPUT), { recursive: true });
  writeFileSync(OUTPUT, `${JSON.stringify({ underlying: UNDERLYING, generatedAt: new Date().toISOString(), spot, seam, seamNote: note, contracts }, null, 2)}\n`);

  logger.info(`Wrote ${contracts.length} contracts — ${calls} calls, ${contracts.length - calls} puts — across ${expirations.size} expirations to ${OUTPUT}.`);
  if (empty > 0) {
    logger.info(`Left out ${empty} that came back with neither greeks nor a quote, which no panel could draw.`);
  }
}

main().catch((err: unknown) => {
  logger.error('Could not write the option chain.', err);
  process.exitCode = 1;
});

File: viz/option_chain.py
"""An option chain in eight panels, from the JSON `option-chain.ts` writes.

    npm run option-chain -w @fleece/playground   # fetches, writes data/option-chain.json
    uv run --project viz viz/option_chain.py     # draws it

Two rows and four columns: calls on top, puts below; delta, implied volatility, theta and
the mid of the quote across. Every panel shares both axes — strike across, expiration up —
so a feature lands at the same place in all eight.

`tricontourf` rather than `contourf` because the chain is ragged: weeklies list far fewer
strikes than the monthlies beside them, and a gridded plotter would have to invent the
missing ones. Triangles bridging the gaps at the edge of the chain are dropped, so the
footprint is the shape of the data rather than its convex hull.

The y axis is the square root of the days to expiry, the axis on which a flat term
structure is flat and a diffusion spreads at a constant angle. The ticks carry dates.

Every contract the writer priced is drawn, both sides of spot. What varies is which
contracts the twelve colours are spent on — the `scale` and `calibrate` fields on `Panel`.
"""

import json
from datetime import date, datetime
from pathlib import Path
from typing import Callable, NamedTuple

import matplotlib.patheffects as effects
import matplotlib.pyplot as plt
import numpy as np
from matplotlib.ticker import FuncFormatter, MaxNLocator
from matplotlib.tri import Triangulation

DATA = Path(__file__).parent / "data" / "option-chain.json"
OUTPUT = Path(__file__).parent / "data" / "option-chain.png"

# Text and furniture stay grey; only the contour fills carry a value.
INK = "#1f2328"
MUTED = "#6b7280"
GRID = "#dcdfe4"
SURFACE = "#ffffff"
SPOT = "#c2410c"

# Viridis is monotonic in lightness and holds that order under every kind of colour
# blindness. Delta gets a diverging pair because zero is a real place on its scale.
SEQUENTIAL = "viridis"
DIVERGING = "RdBu_r"

# Roughly this many bands per panel — roughly, because the boundaries are then moved onto
# round numbers, so a value can be read off the colourbar without interpolating.
LEVELS = 20

# Label every nth contour line. All of them on a panel this size is a thicket.
LABEL_EVERY = 5

# At most this many labels on a colourbar, whatever `LEVELS` is set to. Every band boundary
# is a tick, and past this they collide on a bar this narrow.
MAX_BAR_TICKS = 8

# The percentile the colour scale runs between, so one outlying contract cannot set the
# ends of a ramp the whole chain is drawn on. Where it bites, the colourbar grows an arrow;
# every contract is drawn either way.
ROBUST_PERCENTILE = 2.0

# A triangle whose longest side exceeds this many times the median is dropped: Delaunay
# fills the convex hull, and the slivers reaching across the chain's ragged corners
# interpolate between contracts that have nothing to do with each other. Measured in
# axis-normalised space, so "long" means long as drawn.
SLIVER = 4.0

# Roughly this many dates on the y axis, thinned by how far apart they fall as drawn —
# every nth would keep a colliding pair and drop a well-spaced one.
MAX_DATE_TICKS = 7


class Panel(NamedTuple):
    """One column: the measure it draws, and how its colour scale is built.

    `scale` spaces the bands: `linear` by equal intervals, `ratio` by equal multiples (for
    a measure spread over orders of magnitude — theta goes as one over the root of the time
    left, and on a linear scale 58% of a real chain landed in one band), `diverging` about
    zero for a measure with a sign.

    `calibrate` picks which contracts set the ends — `all`, or `out-of-the-money` for a
    measure of time value, whose in-the-money reading is mostly the width of a quote that
    is nearly all intrinsic. Everything is drawn either way; a contract outside the range
    saturates into the end band under the colourbar's arrow.
    """

    key: str
    title: str
    scale: str
    calibrate: str
    format: Callable[[float], str]


def money(value: float) -> str:
    """Cents below ten dollars, whole dollars above — one panel holds both."""
    return f"${value:,.0f}" if abs(value) >= 10 else f"${value:,.2f}"


def decay(value: float) -> str:
    """Ratio spacing puts 0.006 and 2.4 on one bar, which no fixed precision writes."""
    return f"{value:.2f}" if abs(value) >= 0.1 else f"{value:.3f}"


PANELS = (
    Panel("delta", "Delta", "diverging", "all", lambda v: f"{v:.2f}"),
    Panel("iv", "Implied volatility", "linear", "out-of-the-money", lambda v: f"{v * 100:.0f}%"),
    Panel("theta", "Theta", "ratio", "all", decay),
    # `ratio` here if the wings matter more than the body: premium runs from half a cent to
    # a hundred dollars, and linear spends most of the ramp above $50.
    Panel("mid", "Mid price", "linear", "all", money),
)

ROWS = (("call", "Calls"), ("put", "Puts"))


def main() -> None:
    if not DATA.exists():
        raise SystemExit(f"No data at {DATA}. Run: npm run option-chain -w @fleece/playground")

    payload = json.loads(DATA.read_text())
    contracts = payload["contracts"]
    if not contracts:
        raise SystemExit(f"{DATA} holds no contracts.")

    as_of = datetime.fromisoformat(payload["generatedAt"]).date()
    spot = payload["spot"]
    chains = {side: [c for c in contracts if c["type"] == side] for side, _ in ROWS}

    apply_style()
    figure, grid = plt.subplots(2, 4, figsize=(18, 9), sharex=True, sharey=True, layout="constrained")
    figure.get_layout_engine().set(rect=(0.006, 0.006, 0.988, 0.885))

    for column, panel in enumerate(PANELS):
        levels, extend = contour_levels(chains, panel, spot)
        colours = DIVERGING if panel.scale == "diverging" else SEQUENTIAL
        drawn = None

        for row, (side, _) in enumerate(ROWS):
            axes = grid[row][column]
            drawn = draw(axes, chains[side], panel, levels, extend, colours, as_of) or drawn
            axes.axvline(spot, color=SPOT, linewidth=1.1, linestyle=(0, (4, 3)), zorder=6)

        if drawn is not None:
            bar = figure.colorbar(
                drawn,
                ax=list(grid[:, column]),
                location="bottom",
                shrink=0.86,
                aspect=34,
                pad=0.015,
                format=FuncFormatter(lambda v, _, write=panel.format: write(v)),
                ticks=bar_ticks(levels, panel),
            )
            bar.outline.set_edgecolor(GRID)
            bar.ax.tick_params(colors=MUTED, length=0, labelsize=8)

        grid[0][column].set_title(panel.title, fontsize=11, color=INK, pad=8)

    for row, (_, label) in enumerate(ROWS):
        grid[row][0].set_ylabel(label, fontsize=12, color=INK, weight="bold", labelpad=8)

    label_expirations(grid[0][0], contracts, as_of)
    grid[1][0].set_xlabel("Strike", fontsize=10, color=MUTED, labelpad=6)

    title(figure, payload, contracts, as_of, spot)

    figure.savefig(OUTPUT, dpi=160)
    print(f"Wrote {OUTPUT}")
    plt.show()


def apply_style() -> None:
    plt.rcParams.update(
        {
            "figure.facecolor": SURFACE,
            "axes.facecolor": SURFACE,
            "savefig.facecolor": SURFACE,
            "axes.edgecolor": GRID,
            "axes.linewidth": 0.8,
            "text.color": INK,
            "axes.labelcolor": MUTED,
            "xtick.color": MUTED,
            "ytick.color": MUTED,
            "font.size": 10,
        }
    )


def bar_ticks(levels, panel: Panel):
    """Every nth band boundary, thinned to `MAX_BAR_TICKS`.

    Counted outwards from the centre on a diverging scale, so zero always carries a tick —
    it is the one value on that bar a reader needs to find.
    """
    stride = max(1, -(-len(levels) // MAX_BAR_TICKS))
    if panel.scale != "diverging":
        return levels[::stride]
    middle = (len(levels) - 1) // 2
    return levels[middle % stride :: stride]


def draw(axes, chain, panel: Panel, levels, extend: str, colours, as_of: date):
    """One panel. Returns the filled contour set, or None where there was nothing to draw."""
    axes.tick_params(colors=MUTED, length=0, labelsize=9)

    strike, term, value = points(chain, panel.key, as_of)
    triangles = triangulate(strike, term)
    if triangles is None:
        axes.text(0.5, 0.5, "not enough priced contracts", transform=axes.transAxes, ha="center", va="center", color=MUTED, fontsize=9)
        return None

    filled = axes.tricontourf(triangles, value, levels=levels, cmap=colours, extend=extend)

    # The bands are already edge to edge, so the lines exist only to be labelled. A fill
    # running near-black to near-yellow has no one legible text colour, so each label
    # carries a white halo out with it.
    lines = axes.tricontour(triangles, value, levels=levels[1:-1], colors=INK, linewidths=0.3, alpha=0.16)
    labels = axes.clabel(lines, levels=levels[1:-1][::LABEL_EVERY], fmt=panel.format, fontsize=8, colors=INK, inline_spacing=4)
    for label in labels:
        label.set_path_effects([effects.withStroke(linewidth=1.6, foreground=SURFACE)])

    return filled


def points(chain, key: str, as_of: date):
    """The contracts of one chain that carry this measure, as three parallel arrays."""
    kept = [c for c in chain if c.get(key) is not None]
    strike = np.array([c["strike"] for c in kept], dtype=float)
    term = np.array([term_axis(days_to(c["expiration"], as_of)) for c in kept], dtype=float)
    value = np.array([c[key] for c in kept], dtype=float)
    return strike, term, value


def contour_levels(chains, panel: Panel, spot: float):
    """The band boundaries for one column, and which ends of them the data runs past.

    Shared by both rows, which is the point of the column: volatility on one scale is where
    a stale chain gives itself away, as two rows at visibly different levels for contracts
    parity says share a number.
    """
    drawn = measured(chains, panel.key, lambda _: True, spot)
    if drawn.size == 0:
        return np.linspace(0.0, 1.0, LEVELS + 1), "neither"

    if panel.scale == "diverging":
        bound = float(np.abs(drawn).max()) or 1.0
        return MaxNLocator(nbins=LEVELS, symmetric=True).tick_values(-bound, bound), "neither"

    calibrating = drawn if panel.calibrate == "all" else measured(chains, panel.key, lambda c: out_of_the_money(c, spot), spot)
    if calibrating.size == 0:
        calibrating = drawn

    low, high = (float(v) for v in np.percentile(calibrating, [ROBUST_PERCENTILE, 100 - ROBUST_PERCENTILE]))
    if high <= low:
        low, high = float(calibrating.min()), float(calibrating.max())
    if high <= low:
        return np.linspace(low - 0.5, low + 0.5, LEVELS + 1), "neither"

    levels = ratio_levels(low, high) if panel.scale == "ratio" else None
    if levels is None:
        levels = trim(MaxNLocator(nbins=LEVELS).tick_values(low, high), calibrating)

    return levels, extent_of(drawn, levels)


def measured(chains, key: str, keep, spot: float) -> np.ndarray:
    """Every value of one measure across both chains, from the contracts `keep` accepts."""
    return np.array(
        [c[key] for chain in chains.values() for c in chain if c.get(key) is not None and keep(c)],
        dtype=float,
    )


def out_of_the_money(contract, spot: float) -> bool:
    """A call above spot, a put below. The strike landing on spot goes to the puts."""
    return contract["strike"] > spot if contract["type"] == "call" else contract["strike"] <= spot


def ratio_levels(low: float, high: float):
    """Boundaries spaced by equal multiples, worked out on the magnitudes and signed back.

    `None` where the range straddles zero, which no ratio can span.
    """
    if low > 0 and high > 0:
        return np.geomspace(low, high, LEVELS + 1)
    if low < 0 and high < 0:
        return -np.geomspace(abs(low), abs(high), LEVELS + 1)
    return None


def trim(levels: np.ndarray, values: np.ndarray) -> np.ndarray:
    """The rounded boundaries, with the ones no contract reaches dropped.

    An empty band takes a colour off the ramp, so the outermost contracts get drawn in the
    second colour from the end and the reader hunts the chart for the first.
    """
    first = max(0, int(np.searchsorted(levels, values.min(), side="right")) - 1)
    last = min(levels.size - 1, int(np.searchsorted(levels, values.max(), side="left")))
    return levels[first : last + 1] if last > first else levels


def extent_of(values: np.ndarray, levels) -> str:
    """Which ends of the scale the data runs past, in the word `contourf` wants for it.

    Asked of every contract drawn, not of the ones the scale was calibrated on.
    """
    below, above = float(values.min()) < levels[0], float(values.max()) > levels[-1]
    if below and above:
        return "both"
    return "min" if below else "max" if above else "neither"


def triangulate(strike: np.ndarray, term: np.ndarray):
    """Delaunay over the contracts, with the slivers that bridge the chain's gaps dropped.

    Triangulated in axis-normalised space and rebuilt on the real coordinates: strike runs
    over hundreds of dollars and root-time over about ten, so triangulating the raw numbers
    would judge shape and slivers by the units rather than the data.
    """
    if strike.size < 3:
        return None

    try:
        normalised = Triangulation(unit(strike), unit(term))
    except (ValueError, RuntimeError):
        return None

    triangles = Triangulation(strike, term, triangles=normalised.triangles)
    longest = triangle_sides(unit(strike), unit(term), normalised.triangles).max(axis=1)
    mask = longest > SLIVER * float(np.median(longest))
    if mask.any() and not mask.all():
        triangles.set_mask(mask)
    return triangles


def triangle_sides(x: np.ndarray, y: np.ndarray, triangles: np.ndarray) -> np.ndarray:
    """The three side lengths of every triangle, as a (triangles, 3) array."""
    corners = np.stack([x, y], axis=1)[triangles]
    return np.linalg.norm(corners - np.roll(corners, 1, axis=1), axis=2)


def unit(values: np.ndarray) -> np.ndarray:
    """The values rescaled onto 0 to 1, so two axes in different units can be compared."""
    low, high = float(values.min()), float(values.max())
    return (values - low) / (high - low) if high > low else np.zeros_like(values)


def days_to(expiration: str, as_of: date) -> int:
    return (date.fromisoformat(expiration) - as_of).days


def term_axis(days: float) -> float:
    """Where a number of days to expiry sits on the y axis. See the module docstring."""
    return float(np.sqrt(max(days, 0)))


def label_expirations(axes, contracts, as_of: date) -> None:
    """Ticks at the expirations that exist, labelled with the date rather than the count."""
    expirations = sorted({c["expiration"] for c in contracts})
    positions = [term_axis(days_to(e, as_of)) for e in expirations]
    smallest_gap = (max(positions) - min(positions)) / MAX_DATE_TICKS

    shown = []
    for expiration, position in zip(expirations, positions):
        if not shown or position - shown[-1][1] >= smallest_gap:
            shown.append((expiration, position))

    axes.set_yticks([position for _, position in shown])
    axes.set_yticklabels([f"{date.fromisoformat(e):%-d %b}" for e, _ in shown], fontsize=9)


def title(figure, payload, contracts, as_of: date, spot: float) -> None:
    calls = sum(1 for c in contracts if c["type"] == "call")
    expirations = len({c["expiration"] for c in contracts})

    figure.suptitle(f"{payload['underlying']} option chain", x=0.008, y=0.982, ha="left", fontsize=16, weight="bold", color=INK)
    figure.text(
        0.008,
        0.947,
        f"{calls} calls and {len(contracts) - calls} puts across {expirations} expirations · "
        f"{as_of:%-d %B %Y} · strike across, expiration up · the dashed line is spot at {spot:,.2f}",
        ha="left",
        fontsize=10,
        color=MUTED,
    )

    stamp = seam_stamp(payload.get("seam"))
    if stamp is not None:
        # `wrap` is the safety net: the stamp is written to fit one line at this width.
        figure.text(0.008, 0.917, stamp, ha="left", fontsize=9, color=SPOT, wrap=True)


def seam_stamp(seam) -> str | None:
    """The parity warning, where there is one, in a line."""
    if seam is None:
        return "No strike is quoted on both sides, so put-call parity could not be checked. Read the two rows as unverified."
    if not seam["stale"]:
        return None
    return (
        f"Stale quotes: the calls and puts disagree by {seam['disagreement'] * 100:.1f} volatility points at the money against a "
        f"tolerance of {seam['tolerance'] * 100:.1f} — the greeks are marked against an underlying ${abs(seam['greeksAhead']):.2f} "
        f"{'above' if seam['greeksAhead'] >= 0 else 'below'} the quoted one, so the two rows sit offset. Re-run during the option session."
    )


if __name__ == "__main__":
    main()
